skrifa/src/outline/native_hint.rs

```diff
@@ -0,0 +1,147 @@
+    /// Same as light, but with additional support for RGB subpixel rendering.
+    LightSubpixel,
+    /// Same as light subpixel, but always prevents adjustment in the
+    /// horizontal direction. This is the default mode.
+    VerticalSubpixel,
+}
+
+/// Hinting instance that uses information embedded in the font to perform
+/// grid-fitting.
+#[derive(Clone)]
+pub struct NativeHintingInstance {
+    size: Size,
+    coords: Vec<NormalizedCoord>,
+    mode: NativeHinting,
+    kind: HinterKind,
+}
+
+impl NativeHintingInstance {
+    /// Creates a new native hinting instance for the given outline
+    /// collection, size, location in variation space and hinting mode.
+    pub fn new<'a>(
+        outline_glyphs: &OutlineGlyphCollection,
+        size: Size,
+        location: impl Into<LocationRef<'a>>,
+        mode: NativeHinting,
+    ) -> Result<Self, DrawError> {
+        let mut hinter = Self {
+            size: Size::unscaled(),
+            coords: vec![],
+            mode,
+            kind: HinterKind::None,
+        };
+        hinter.reconfigure(outline_glyphs, size, location, mode)?;
+        Ok(hinter)
+    }
+
+    /// Returns the currently configured size.
+    pub fn size(&self) -> Size {
+        self.size
+    }
+
+    /// Returns the currently configured normalized location in variation space.
+    pub fn location(&self) -> LocationRef {
+        LocationRef::new(&self.coords)
+    }
+
+    /// Returns the currently configured hinting mode.
+    pub fn mode(&self) -> NativeHinting {
+        self.mode
+    }
+
+    /// Resets the hinter state for a new font instance with the given
+    /// outline collection and settings.
+    pub fn reconfigure<'a>(
+        &mut self,
+        outlines: &OutlineGlyphCollection,
+        size: Size,
+        location: impl Into<LocationRef<'a>>,
+        mode: NativeHinting,
+    ) -> Result<(), DrawError> {
+        self.size = size;
+        self.coords.clear();
+        self.coords.extend_from_slice(location.into().coords());
+        self.mode = mode;
+        // Reuse memory if the font contains the same outline format
+        let current_kind = core::mem::replace(&mut self.kind, HinterKind::None);
+        match &outlines.kind {
+            OutlineCollectionKind::Glyf(_) => {
+                self.kind = HinterKind::Glyf();
+            }
+            OutlineCollectionKind::Cff(cff) => {
+                let mut subfonts = match current_kind {
+                    HinterKind::Cff(subfonts) => subfonts,
+                    _ => vec![],
+                };
+                subfonts.clear();
+                let size = size.ppem().unwrap_or_default();
+                for i in 0..cff.subfont_count() {
+                    subfonts.push(cff.subfont(i, size, &self.coords)?);
+                }
+                self.kind = HinterKind::Cff(subfonts);
+            }
+            OutlineCollectionKind::None => {}
+        }
+        Ok(())
+    }
+
+    pub(super) fn draw_hinted(
```

Comment:
Optional: draw_hinted _on a hinter feels a bit redundant, it's not like you'd ask NativeHintingInstance to draw any other way is it?

Reply:
Hm. I guess with HinterKind - which feels like maybe it belongs up by the NativeHinting enum - you could have None? I don't find it obvious when you would want to do that though...?

Reply:
Happy to drop the _hinted. 

The None variant exists because we can potentially have a font with no outline tables and this state represents that. I can make initialization return an error in this case if you like that better.

Reply:
draw_hinted -> draw and added a comment explaining why the `None` variant exists